### PR TITLE
[camera] Convert iOS Obj-C->Dart calls to Pigeon

### PR DESCRIPTION
**Pull request by stuartmorgan**
Converts all of the Obj-C -> Dart calls to Pigeon, using the new suffix-based Pigeon API instantiation feature.

This required decentralizing some threading code slightly: since method channel calls only involve one method (due to the lack of strong types), a wrapper that automatically did thread bouncing was feasible, but with Pigeon it's not since a wrapper would have to duplicate the entire API surface, and that's more work than just doing the dispatches at the call site.

Part of https://github.com/flutter/flutter/issues/117905

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_avfoundation/ios/Classes/CameraProperties.h [hunk 1]
```diff
@@ -30,49 +32,24 @@ extern AVCaptureFlashMode FLTGetAVCaptureFlashModeForFLTFlashMode(FLTFlashMode m
 
 #pragma mark - exposure mode
 
-/// Represents camera's exposure mode. Mirrors ExposureMode in camera.dart.
-typedef NS_ENUM(NSInteger, FLTExposureMode) {
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/ios/Classes/CameraProperties.h:34]: These enums are no longer needed because we can just use the Pigeon-generated enums instead.

File: packages/camera/camera_avfoundation/ios/Classes/CameraProperties.h [hunk 2]
```diff
@@ -30,49 +32,24 @@ extern AVCaptureFlashMode FLTGetAVCaptureFlashModeForFLTFlashMode(FLTFlashMode m
...
-/// Gets a string representation of exposure mode.
-/// @param mode exposure mode
-extern NSString *FLTGetStringForFLTExposureMode(FLTExposureMode mode);
-
-/// Gets FLTExposureMode from its string representation.
-/// @param mode a string representation of the FLTExposureMode.
-extern FLTExposureMode FLTGetFLTExposureModeForString(NSString *mode);
+/// Gets FCPPlatformExposureMode from its string representation.
+/// @param mode a string representation of the exposure mode.
+extern FCPPlatformExposureMode FCPGetExposureModeForString(NSString *mode);
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/ios/Classes/CameraProperties.h:37]: These are temporary methods; there are Dart->host calls that send exposure and focus modes, so until I convert those in a follow-up we need to be able to convert method channel strings to the Pigeon enum that replaced the ones removed here.

File: packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m [hunk 3]
```diff
@@ -169,25 +174,10 @@ - (void)handleMethodCallAsync:(FlutterMethodCall *)call result:(FlutterResult)re
...
-               @"exposureMode" : FLTGetStringForFLTExposureMode([_camera exposureMode]),
-               @"focusMode" : FLTGetStringForFLTFocusMode([_camera focusMode]),
-               @"exposurePointSupported" :
-                   @([_camera.captureDevice isExposurePointOfInterestSupported]),
-               @"focusPointSupported" : @([_camera.captureDevice isFocusPointOfInterestSupported]),
-             }];
+      _camera.dartAPI = [[FCPCameraEventApi alloc]
+          initWithBinaryMessenger:_messenger
+             messageChannelSuffix:[NSString stringWithFormat:@"%ld", cameraId]];
+      [_camera reportInitializationState];
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m:180]: I moved the actual channel invocation into `FLTCam` because it seemed strange to have it here when a) the camera instance already has the channel, and b) all of the values here come from the camera instance.

File: packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m [hunk 4]
```diff
@@ -41,37 +41,34 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
...
-            binaryMessenger:_messenger];
-  _deviceEventMethodChannel =
-      [[FLTThreadSafeMethodChannel alloc] initWithMethodChannel:methodChannel];
-}
-
-- (void)detachFromEngineForRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
-  [UIDevice.currentDevice endGeneratingDeviceOrientationNotifications];
-}
-
-- (void)startOrientationListener {
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m:69]: I inlined this as a drive-by cleanup because it was only being called from one place, and that was `init`, which shouldn't be calling methods.

File: packages/camera/camera_avfoundation/ios/Classes/FLTCam.m [hunk 5]
```diff
@@ -959,17 +979,7 @@ - (void)setFlashModeWithResult:(FlutterResult)result mode:(NSString *)modeStr {
...
-  if (mode == FLTExposureModeInvalid) {
-    result(FlutterErrorFromNSError([NSError
-        errorWithDomain:NSCocoaErrorDomain
-                   code:NSURLErrorUnknown
-               userInfo:@{
-                 NSLocalizedDescriptionKey :
-                     [NSString stringWithFormat:@"Unknown exposure mode %@", modeStr]
-               }]));
-    return;
-  }
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/ios/Classes/FLTCam.m:972]: I'm now preventing this error from being possible on the Dart side (same below).

File: packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart [hunk 6]
```diff
@@ -552,6 +536,38 @@ class AVFoundationCamera extends CameraPlatform {
     return Texture(textureId: cameraId);
   }
 
+  String _serializeFocusMode(FocusMode mode) {
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/lib/src/avfoundation_camera.dart:539]: I added these instead of continuing to use the ones from the platform interface package (the versions without the `_`) because having serialization in the platform interface package and deserialization in the implementation package is an anti-pattern since it means changes in one package can break internal communication in another. We created a lot of it by accident during the initial process of federating plugins; Pigeon migration has been eliminating it, but doing it now meant I could remove all the native code that required an "invalid" state for these enums, since I can guarantee within this package that we won't get a state at the native layer that we don't handle.

(These methods will become enum->enum conversion methods instead of enum->string conversion when I finish the Pigeon migration for this package.)

File: packages/camera/camera_avfoundation/test/avfoundation_camera_test.dart [hunk 7]
```diff
@@ -361,77 +343,27 @@ void main() {
...
-          MethodCall('initialized', event.toJson()), cameraId);
-
-      // Assert
-      expect(await streamQueue.next, event);
-
-      // Clean up
-      await streamQueue.cancel();
-    });
-
-    test('Should receive resolution changes', () async {
```
Comment (stuartmorgan) [inline packages/camera/camera_avfoundation/test/avfoundation_camera_test.dart:384]: This and the next test are testing functionality that was dead code. The Dart method call handler handled several messages that are never sent by the native code. It doesn't look like Android uses them either, so I'm guessing it was code mistakenly left in during refactoring in code review of a feature.

File: packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m [hunk 8]
```diff
@@ -93,9 +90,17 @@ - (void)orientationChanged:(NSNotification *)note {
 }
 
 - (void)sendDeviceOrientation:(UIDeviceOrientation)orientation {
-  [_deviceEventMethodChannel
-      invokeMethod:@"orientation_changed"
-         arguments:@{@"orientation" : FLTGetStringForUIDeviceOrientation(orientation)}];
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [weakSelf.globalEventAPI
```
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m:95]: I vaguely remember there was an effort to make method channel thread safe. what's the progress on that and do we still need to manually dispatch to main in plugin code?

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraOrientationTests.m [hunk 9]
```diff
@@ -9,44 +9,87 @@
...
+#pragma mark -
+
 @interface CameraOrientationTests : XCTestCase
 @end
 
 @implementation CameraOrientationTests
 
+// Ensure that the given queue and then the main queue have both cycled, to wait for any pending
+// async events that may have been bounced between them.
+- (void)waitForRoundTripWithQueue:(dispatch_queue_t)queue {
```
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/example/ios/RunnerTests/CameraOrientationTests.m:42]: nice

File: packages/camera/camera_avfoundation/ios/Classes/CameraProperties.m [hunk 10]
```diff
@@ -36,51 +36,29 @@ AVCaptureFlashMode FLTGetAVCaptureFlashModeForFLTFlashMode(FLTFlashMode mode) {
...
-    case FLTFocusModeInvalid:
-      // This value should never actually be used.
-      return nil;
-  }
-  return nil;
-}
-
-FLTFocusMode FLTGetFLTFocusModeForString(NSString *mode) {
+FCPPlatformFocusMode FCPGetFocusModeForString(NSString *mode) {
   if ([mode isEqualToString:@"auto"]) {
```
Comment (hellohuanlin) [inline packages/camera/camera_avfoundation/ios/Classes/CameraProperties.m:54]: will these hard coded strings be gone eventually after pigeon work is done?

**Review by hellohuanlin — APPROVED** [review]
Went through a thorough review. Looks pretty clean.

File: packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m (hunk 8 above)
Comment (stuartmorgan) [reply packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m:95]: The current state is that replies are thread-safe, but calls that start on the host side still have to be done from the main thread. I don't think there's active work on addressing that part (although I believe it's still something we do eventually want to do).

File: packages/camera/camera_avfoundation/ios/Classes/CameraProperties.m (hunk 10 above)
Comment (stuartmorgan) [reply packages/camera/camera_avfoundation/ios/Classes/CameraProperties.m:54]: Yes, this is just for the remaining non-Pigeon Dart->host calls.
